core/drivers/crypto/caam/caam_key.c

```diff
@@ -0,0 +1,753 @@
+	}
+
+err:
+	caam_key_free(out_key);
+out:
+	caam_free_desc(&desc);
+	return status;
+}
+
+enum caam_status caam_key_deserialize_from_bin(uint8_t *data, size_t size,
+					       struct caamkey *key,
+					       size_t sec_size)
+{
+	enum caam_status status = CAAM_FAILURE;
+	struct caamkey blob = { };
+
+	assert(data && size && key);
+
+	KEY_TRACE("Deserialization binary buffer");
+	KEY_DUMPBUF("Deserialize key buffer input", data, size);
+
+	/*
+	 * If a security key size is given, use it. Otherwise, rely on
+	 * the buffer size.
+	 * In some case, like ECC keys, the bignum size is less than the
+	 * security size and it requires the key to be padded with 0's.
+	 */
+	if (sec_size == 0)
+		sec_size = get_key_sec_size(data, size);
+
+	blob.key_type = get_key_type(data, size);
+	blob.sec_size = sec_size;
+	blob.is_blob = true;
+
+	if (blob.key_type == CAAM_KEY_PLAIN_TEXT) {
+		key->sec_size = blob.sec_size;
+		key->key_type = blob.key_type;
+		key->is_blob = false;
+
+		status = caam_key_alloc(key);
+		if (status) {
+			KEY_TRACE("Key allocation error");
+			return status;
+		}
+
+		/* Some asymmetric keys have leading zeros we must preserve */
+		memcpy(key->buf.data + key->buf.length - size, data, size);
+
+		return CAAM_NO_ERROR;
+	}
+
+	status = caam_key_alloc(&blob);
+	if (status) {
+		KEY_TRACE("Key allocation error");
+		return status;
+	}
+
+	memcpy(blob.buf.data, data + get_key_buf_offset(data, size),
+	       get_key_buf_size(data, size));
+
+	/* Set destination key */
+	key->key_type = blob.key_type;
+	key->sec_size = blob.sec_size;
+	key->is_blob = false;
+
+	/* De-blob operation */
+	status = caam_key_operation_blob(&blob, key);
+	if (status) {
+		KEY_TRACE("De-blob operation fail");
+		goto out;
+	}
+
+	KEY_TRACE("Deserialization binary buffer done");
+	caam_key_dump("Deserialization output key", key);
+out:
+	caam_key_free(&blob);
+	return status;
+}
+
+enum caam_status caam_key_serialize_to_bin(uint8_t *data, size_t size,
+					   const struct caamkey *key)
+{
+	struct caam_key_serialized key_ser = { };
+	struct caamkey blob = { };
+	enum caam_status status = CAAM_FAILURE;
+	size_t serialized_size = 0;
+
+	assert(data && size && key);
+
+	caam_key_dump("Serialization input key", key);
+
+	/* If the key is plain text, just copy key to buffer */
+	if (key->key_type == CAAM_KEY_PLAIN_TEXT) {
+		if (size < key->buf.length) {
+			KEY_TRACE("Buffer is too short");
+			return CAAM_SHORT_BUFFER;
+		}
+
+		memcpy(data, key->buf.data, key->buf.length);
+
+		return CAAM_NO_ERROR;
+	}
+
+	/* The input key must not be a blob */
+	assert(!key->is_blob);
+
+	/* Blob the given key for serialization and export */
+	blob.is_blob = true;
+	blob.sec_size = key->sec_size;
+	blob.key_type = key->key_type;
+
+	/*
+	 * Check if the destination is big enough for the black blob buffer and
+	 * header.
+	 */
+	status = caam_key_serialized_size(&blob, &serialized_size);
+	if (status)
+		goto out;
```

Comment:
Nothing is allocated yet, why not just return the error?

Reply:
Let's return the error indeed